sound/soc/sof/debug.c

```diff
@@ -17,6 +17,142 @@
 #include "sof-priv.h"
 #include "ops.h"
 
+static ssize_t sof_dfsentry_write(struct file *file, const char __user *buffer,
+				  size_t count, loff_t *ppos)
+{
+	struct snd_sof_dfsentry *dfse = file->private_data;
+	struct snd_sof_dev *sdev = dfse->sdev;
+	struct sof_ipc_cmd_hdr hdr;
+	struct sof_ipc_reply reply;
+	size_t size;
+	char *string;
+	int ret;
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_ENABLE_IPC_FLOOD_TEST)
+	ktime_t start, end, test_end;
+	s64 ipc_response_time;
+	s64 avg_response_time = 0;
+	s64 min_response_time = LONG_MAX;
+	s64 max_response_time = 0;
+	int i = 0;
+	int err;
+	unsigned long ipc_count = 0;
+	unsigned long ipc_duration_ms = 0;
+	bool flood_duration_test = false;
+#endif
+	string = kzalloc(count, GFP_KERNEL);
+	if (!string)
+		return -ENOMEM;
+
+	size = simple_write_to_buffer(string, count, ppos, buffer, count);
+	ret = size;
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_ENABLE_IPC_FLOOD_TEST)
+	if (!strcmp(dfse->dfsentry->d_name.name, "ipc_flood_count"))
+		goto test;
+
+	if (!strcmp(dfse->dfsentry->d_name.name, "ipc_flood_duration_ms")) {
+		flood_duration_test = true;
+		goto test;
+	}
+
+	/*
+	 * write op is only supported for ipc_flood_count or
+	 * ipc_flood_duration_ms debugfs entries atm.
+	 * ipc_flood_count floods the DSP with the number of IPC's specified.
+	 * ipc_duration_ms test floods the DSP for the time specified
+	 * in the debugfs entry.
+	 */
+	return -EINVAL;
+test:
```

Comment:
Can you move the test logic to a separate function? This function starts to be waaayyy too long...

Reply:
agree

Reply:
mmmh.... May I suggest getting rid of gotos?
```
if (!strcmp(dfse->dfsentry->d_name.name, "ipc_flood_duration_ms"))
	flood_duration_test = true;
else if (strcmp(dfse->dfsentry->d_name.name, "ipc_flood_duration_ms"))
	return -EINVAL;
```

Reply:
agree, this is not so nice.